File: Site/frontend/src/types/card.ts
/**
 * Types pour le système de cartes Citadelle
 * Correspond exactement au système du bot Discord
 */

// Catégories de cartes (ordre de rareté décroissante)
export const ALL_CATEGORIES = [
  'Secrète',
  'Fondateur',
  'Historique',
  'Maître',
  'Black Hole',
  'Architectes',
  'Professeurs',
  'Autre',
  'Élèves',
] as const

export type CardCategory = (typeof ALL_CATEGORIES)[number]

// Poids de rareté (probabilités de tirage)
export const RARITY_WEIGHTS: Record<CardCategory, number> = {
  Secrète: 0.005,
  Fondateur: 0.01,
  Historique: 0.02,
  Maître: 0.06,
  'Black Hole': 0.06,
  Architectes: 0.07,
  Professeurs: 0.1167,
  Autre: 0.2569,
  Élèves: 0.4203,
}

// Couleurs par catégorie pour l'UI
export const CATEGORY_COLORS: Record<CardCategory, { bg: string; border: string; text: string }> = {
  Secrète: {
    bg: 'from-purple-900 via-pink-900 to-purple-900',
    border: 'border-purple-500',
    text: 'text-purple-300',
  },
  Fondateur: {
    bg: 'from-yellow-900 via-amber-900 to-yellow-900',
    border: 'border-yellow-500',
    text: 'text-yellow-300',
  },
  Historique: {
    bg: 'from-blue-900 via-indigo-900 to-blue-900',
    border: 'border-blue-500',
    text: 'text-blue-300',
  },
  Maître: {
    bg: 'from-red-900 via-rose-900 to-red-900',
    border: 'border-red-500',
    text: 'text-red-300',
  },
  'Black Hole': {
    bg: 'from-gray-900 via-slate-900 to-gray-900',
    border: 'border-gray-500',
    text: 'text-gray-300',
  },
  Architectes: {
    bg: 'from-cyan-900 via-teal-900 to-cyan-900',
    border: 'border-cyan-500',
    text: 'text-cyan-300',
  },
  Professeurs: {
    bg: 'from-green-900 via-emerald-900 to-green-900',
    border: 'border-green-500',
    text: 'text-green-300',
  },
  Autre: {
    bg: 'from-orange-900 via-amber-900 to-orange-900',
    border: 'border-orange-500',
    text: 'text-orange-300',
  },
  Élèves: {
    bg: 'from-slate-800 via-gray-800 to-slate-800',
    border: 'border-slate-500',
    text: 'text-slate-300',
  },
}

// Interface pour une carte
export interface Card {
  category: CardCategory
  name: string // Nom sans .png
  file_id: string | null
  is_full: boolean
  rarity_weight: number
}

// Interface pour une carte dans une collection utilisateur
export interface CardInCollection extends Card {
  count: number // Nombre d'exemplaires possédés
}

// Informations sur une catégorie
export interface CategoryInfo {
  category: CardCategory
  weight: number
  total_cards: number
  percentage: number
}

File: Site/backend/app/models/card.py
"""
Modèles Pydantic pour les cartes.
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class CardPair(BaseModel):
    """Simple paire catégorie/nom pour identifier une carte."""
    category: str = Field(..., description="Catégorie de la carte")
    name: str = Field(..., description="Nom de la carte (sans .png)")


class CardBase(BaseModel):
    """Modèle de base pour une carte."""
    category: str = Field(..., description="Catégorie de la carte")
    name: str = Field(..., description="Nom de la carte (sans .png)")
    file_id: Optional[str] = Field(None, description="ID du fichier Google Drive")
    is_full: bool = Field(False, description="Est-ce une carte Full?")


class Card(CardBase):
    """Modèle complet d'une carte avec métadonnées."""
    rarity_weight: float = Field(..., description="Poids de rareté de la catégorie")
    discoverer_id: Optional[int] = Field(None, description="ID du premier découvreur")
    discoverer_name: Optional[str] = Field(None, description="Nom du premier découvreur")
    discovery_date: Optional[datetime] = Field(None, description="Date de première découverte")


class CardInCollection(CardBase):
    """Carte dans la collection d'un utilisateur avec le nombre d'exemplaires."""
    count: int = Field(1, description="Nombre d'exemplaires possédés", ge=1)
    acquired_date: Optional[str] = Field(None, description="Date d'acquisition")


class UserCollection(BaseModel):
    """Collection de cartes d'un utilisateur."""
    user_id: int
    cards: List[CardInCollection] = Field(default_factory=list)
    total_cards: int = Field(0, description="Nombre total de cartes (avec doublons)")
    unique_cards: int = Field(0, description="Nombre de cartes uniques")
    completion_percentage: float = Field(0.0, description="Pourcentage de complétion", ge=0, le=100)

    class Config:
        json_schema_extra = {
            "example": {
                "user_id": 123456789,
                "cards": [
                    {"category": "Élèves", "name": "Aria", "count": 3},
                    {"category": "Secrète", "name": "Mystère", "count": 1, "is_full": True}
                ],
                "total_cards": 4,
                "unique_cards": 2,
                "completion_percentage": 15.5
            }
        }


class CardDiscovery(BaseModel):
    """Informations sur la decouverte d'une carte."""
    category: str
    name: str
    discoverer_id: int
    discoverer_name: str
    timestamp: str = Field(..., description="Date de decouverte (format string)")
    discovery_index: int = Field(0, description="Numero de decouverte pour cette carte")


class RarityInfo(BaseModel):
    """Informations sur une catégorie de rareté."""
    category: str
    weight: float = Field(..., description="Poids de rareté (probabilité)")
    total_cards: int = Field(..., description="Nombre total de cartes dans cette catégorie")
    percentage: float = Field(..., description="Pourcentage de cette catégorie", ge=0, le=100)

class UpgradedCard(BaseModel):
    """Carte Full obtenue par upgrade (5 cartes normales → 1 Full)."""
    category: str
    name: str = Field(..., description="Nom de la carte Full (avec '(Full)')")
    file_id: Optional[str] = None
    original_name: str = Field(..., description="Nom de la carte normale d'origine")
    sacrificed_count: int = Field(5, description="Nombre de cartes sacrifiées")


class DrawResult(BaseModel):
    """Résultat d'un tirage avec les cartes obtenues et les upgrades éventuels."""
    drawn_cards: List[Card] = Field(..., description="Cartes obtenues par le tirage")
    upgraded_cards: List[UpgradedCard] = Field(
        default_factory=list,
        description="Cartes Full obtenues par conversion automatique (5 normales → 1 Full)"
    )

